### ISSUE TITLE
Training the same model on the same data yielding extremely different test accuracy

### ISSUE THREAD

**Issue by Mirandatz** [tensorflow/tensorflow#56312]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

2.8.0 (From Google Colab)

### Custom Code

No

### OS Platform and Distribution

Google Colab

### CUDA/cuDNN version

Google Colab

### Current Behaviour?

```shell
I wrote a function to load datasets (train, val, and test), create a model, train the model on the dataset train partition, and compute its accuracy on the test partition.

I expected that running this function multiple times in a row would yield similar results, but they are extremely different, ranging from .20 to .70 accuracy.

I tried to remove most sources of randomness, such as using the same numpy rng seed, TensorFlow rng seed, resetting the TensorFlow backend between runs, and using the same seed for shuffling the dataset, etc. The issue persists.

The issue occurs both in my local machine running Tensorflow 2.10.0-dev20220424 and in the Google Colab running TensorFlow 2.8.
```

### Standalone code to reproduce the issue

```shell
I created a Google Colab notebook that reproduces the issue: 

https://colab.research.google.com/drive/1b2DZaAIjqqXDlQ_uJyinVxOn1z_cbAoT?usp=sharing

The dataset I am using is just CIFAR-10 with the training partition split into train and validation:

https://drive.google.com/file/d/1UXJOv9MJ7bu_0woiVVp0BmcnfDzo0hpf/view?usp=sharing
```

### Relevant log output

_No response_</details>

**Comment by tilakrayal** [comment]
Hi @Mirandatz,
I was able to execute the code without any issues and also the accuracy is also the same for every run. Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/6ad3790641f539bedd86ab5338617585/2-8-56312.ipynb). 

Also, setting the seeds is not enough. If you are working on a small dataset you have to switch from GPU to CPU, as the first one has a non-deterministic order of computations.

Moreover, you should get rid of any parallelism in feeding data, as that can also cause problems. Thank you!

**Comment by Mirandatz** [comment]
Thank you for the quick feedback @tilakrayal I will run your notebook soon, but did you manage to run mine using a GPU?

I ask because I understand that there will be some differences when we use GPUs due to differences in execution order, how the numerical errors accumulate, etc. But I did not expect such extreme differences. Allow me to clarify what I mean by "extreme differences":

```
run 3 test accuracy: 0.5475000143051147
run 4 test accuracy: **0.3677000105381012**
```

The difference is as large as 50% between runs 4 and 3. That doesn't look reasonable. I expected differences <10%.

Also, I believe I have found two things that appear to cause (or influence) this inconsistency (when using GPUs): the shuffling (using buffer sizes smaller than the entire dataset) and batching of the data (dropping remainders). I will write a new notebook and post it here ASAP.

**Comment by Mirandatz** [comment]
Using the following parameters to shuffle and batch makes the issue disappear, but I don't know why:

```
dataset.shuffle(buffer_size=dataset.cardinality().numpy(),
                reshuffle_each_iteration=False,
                seed=rng_seed)
        .batch(batch_size=batch_size,
               drop_remainder=True)
```

**Comment by Mirandatz** [comment]
@tilakrayal I tried to run your notebook, but it errors out.

I created a copy of mine and re-ran it, again using a GPU runtime.
https://colab.research.google.com/drive/1OvUkM_Mf1eUwY3LDRUp6Tk5l8alRi9Oa?usp=sharing

Issue persists.
Had accuracies as low as 0.10 and as high as 0.55 :( theres something really weird going on here.

**Comment by tilakrayal** [comment]
@gowthamkpr,
I was able to reproduce the issue on tensorflow-gpu [v2.8](https://colab.research.google.com/gist/tilakrayal/41ccd32091cfba1720c93fc30cbf306f/2-8_gpu.ipynb) and v2.9. In tensorflow-cpu [v2.8](https://colab.research.google.com/gist/tilakrayal/6ad3790641f539bedd86ab5338617585/2-8-56312.ipynb), i was not able to find the issue. Please find the gist.

**Comment by gowthamkpr** [comment]
As mentioned [here](https://stackoverflow.com/questions/72450109/training-the-same-model-on-the-same-data-yielding-extremely-different-test-accur), these are the reasons for the variance in test accuracy and how can you work around it:

Use tf.keras.utils.set_random_seed(some_seed) control all random behavior is needed

Optimize the algorithm as much as possible to mitigate the impact of calculation accuracy. That is, let the algorithm not produce or accumulate small errors in accuracy. There are many adjustable keys, not only the optimizer or the optimizer's parameters, but also other stuff such as bacth_size, gradient_clip, gradient_norm etc.

Use tf.config.experimental.enable_op_determinism if you want to get completely consistent reproducible results and accept its side effects(deceleration).

**Comment by Mirandatz** [comment]
@gowthamkpr Yes, my mistake. I did not realize that test accuracy varying between 0.1 to 0.55, using the same model on the same dataset, was not only acceptable but expected.
